### PR TITLE
Various improvements (Logger class, test additions & fixes)

### PR DESCRIPTION
**Pull request by ljacqu**
I spent a good portion of today afternoon on PWI and made various changes. Haven't tested in-game yet – I'll still do that. Main changes:

- Upgrade Mockito version, yet again. (And yet again, changes in packages)
- Create PwiLogger class to handle all the logging -> fixes the need of injecting the plugin's main class for logging
- Check the required permission for a command centrally instead of doing it individually in each command
- Reduce duplication in building the player's data folder in FileWriter; sometimes the same folder was taken from PWIPlayer (method now removed), and sometimes FileWriter built it itself

Take your time to look at the changes—as always, questions are welcome and you're not forced to merge. :wink:

### CODE REVIEW

**Comment by Gnat008** [comment]
Holy cow you really impress me! lol. Thanks a ton as usual for all these changes, they're about all things I've wanted to do for some time, but haven't had time to do it myself.

I've got some testing to do with 1.11 tomorrow already, I'll add all this to it.

**Comment by ljacqu** [comment]
Thank you :) If I still have time to sneak in a commit this evening (= within the next 11 hours), I can address the todo message I snuck in :P

**Comment by Gnat008** [comment]
Okay, cool. I'll leave it open for that. Also I think there was a typo in one of the log messages, lemme try to find it again... xD

File: src/main/java/me/gnat008/perworldinventory/data/FileWriter.java [hunk 1]
```diff
@@ -112,14 +109,14 @@ public void writeData(final File file, final String data) {
...
-            ex.printStackTrace();
+            PwiLogger.severe("Could not write data to file '" + file + "':", ex);
         } finally {
             try {
                 if (writer != null) {
                     writer.close();
                 }
             } catch (IOException ex) {
-                ex.printStackTrace();
+                PwiLogger.warning("Failed to clease writer for '" + file + "':", ex);
```
Comment (Gnat008) [inline src/main/java/me/gnat008/perworldinventory/data/FileWriter.java]: I'm guessing 'clease' should be 'close'?

Comment (Gnat008) [inline src/main/java/me/gnat008/perworldinventory/data/FileWriter.java]: xD

**Comment by ljacqu** [comment]
This would conclude my PR. Sorry for the double commit ("Create PWIPlayerFactory" with subsequent " 	Make PWIPlayerFactory retrieve economy information itself"). It's been a long day xP

**Comment by Gnat008** [comment]
lol No worries. I hadn't before considered using a factory for PWIPlayers before. It certainly looks cleaner with the economy stuff.

**Comment by ljacqu** [comment]
I'm not a factory fan but I'd say this is one of the cases where it makes sense, especially since it keeps PWIPlayer from being aware of any third-party plugins. I looked at my first commit and saw the "import Vault.Economy" line in PWIPlayer and was like "nope" :tongue:

What might be nice in a later iteration is to make all the economy stuff live inside its own manager class. It would internally check `isEconomyAvailable`, instead of scattering this check in all classes that want to make use of economy. If you find that overkill, I blame writing [enterprise software](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition) 8 hours a day :P

**Comment by Gnat008** [comment]
Yeah PWIPlayer being only self-aware is definitely better.

As a fan of thedailywtf.com and forum member on what.thedailywtf.com, I've heard quite a bit about enterprise software. hehe

**Pull request merged**